### ISSUE TITLE
PostInstall failure occasionally with a 502 (Bad Gateway)

### ISSUE THREAD

**Issue by jasongornall-dd** [cypress-io/cypress#27316]
### Current behavior

Occasional 502 from postInstall script. Seems https://download.cypress.io/desktop/12.14.0?platform=linux&arch=x64 sometimes is a Bad Gateway

### Desired behavior

It should not fail postInstall

### Test code to reproduce

Testing seems to be once a day across 30 to 40 commits

### Cypress Version

12.14.0

### Node version

v16.15.0

### Operating System

Linux

### Debug Logs

```shell
16:04:40  .../cypress@12.14.0/node_modules/cypress postinstall: URL: https://download.cypress.io/desktop/12.14.0?platform=linux&arch=x64
16:04:40  .../cypress@12.14.0/node_modules/cypress postinstall: Error: Failed downloading the Cypress binary.
16:04:40  .../cypress@12.14.0/node_modules/cypress postinstall: Response code: 502
16:04:40  .../cypress@12.14.0/node_modules/cypress postinstall: Response message: Bad Gateway
```

**Comment by MikeMcC399** [comment]
@jasongornall-dd 

I have also seen this error occurring lately running on GitHub CI.
Example https://github.com/cypress-io/cypress-example-kitchensink/actions/runs/5572030102/jobs/10177652188

```text
Mon, 17 Jul 2023 05:17:09 GMT
npm ERR! URL: https://download.cypress.io/desktop/12.17.1?platform=linux&arch=x64
Mon, 17 Jul 2023 05:17:09 GMT
npm ERR! Error: Failed downloading the Cypress binary.
Mon, 17 Jul 2023 05:17:09 GMT
npm ERR! Response code: 502
Mon, 17 Jul 2023 05:17:09 GMT
npm ERR! Response message: Bad Gateway
```

What system were your logs taken from? Was it also GitHub? When was the last time you saw the error?

If this is intermittent and only happening on GitHub then it may be a GitHub network / system error.

**Comment by sneridagh** [comment]
@cypress-github-team-services-sa 

Guys, put your shit together... fails the 20% of the time the last three days.

<img width="1108" alt="image" src="https://github.com/cypress-io/cypress/assets/486927/1b74b692-5d38-4d0e-97e5-d7a1a988fe21">

**Comment by MikeMcC399** [comment]
A PR in https://github.com/cypress-io/github-action was merged just now which contains more than 100 jobs, all of which successfully installed Cypress.

Is anybody still seeing the 502 Bad Gateway error?

If yes, what is the latest timestamp showing this problem?

**Comment by MikeMcC399** [comment]
https://www.cloudflarestatus.com/#past-incidents is showing some related incidents during the past days.

**Comment by nagash77** [comment]
Hi @jasongornall-dd, @sneridagh and @MikeMcC399 I have raised these problems with our infrastructure teams.  They have been doing some migrations for our download services to use a new service.  They are investigating further and have a couple of ideas about ways to limit these types of occurrences.

**Comment by nagash77** [comment]
Hello everyone on this issue, I just wanted to circle back and let everyone know that our infrastructure team has pushed some changes that should address this problem.  I want to apologize again for the inconvenience and we really do appreciate you all bringing this to our attention.  Thank you for being a great community!  I am going to mark this ticket as resolved.  If folks are still seeing this problem going for, please comment back here and we can reopen and investigate.

**Comment by sneridagh** [comment]
@nagash77 Thanks!

**Comment by bradleyayers** [comment]
We're still hitting this occasionally, I just hit it now a couple of times on different build agents:

```
[2023-08-29T09:19:14Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT The Cypress App could not be downloaded.
[2023-08-29T09:19:14Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT
[2023-08-29T09:19:14Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT Does your workplace require a proxy to be used to access the Internet? If so, you must configure the HTTP_PROXY environment variable before downloading Cypress. Read more: https://on.cypress.io/proxy-configuration
[2023-08-29T09:19:14Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT
[2023-08-29T09:19:14Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT Otherwise, please check network connectivity and try again:
[2023-08-29T09:19:14Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT
[2023-08-29T09:19:14Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT ----------
[2023-08-29T09:19:14Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT
[2023-08-29T09:19:14Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT URL: https://download.cypress.io/desktop/10.10.0?platform=linux&arch=x64
[2023-08-29T09:19:14Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT Error: Failed downloading the Cypress binary.
[2023-08-29T09:19:14Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT Response code: 502
[2023-08-29T09:19:14Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT Response message: Bad Gateway
[2023-08-29T09:19:14Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT
[2023-08-29T09:19:14Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT ----------
[2023-08-29T09:19:14Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT
[2023-08-29T09:19:14Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT Platform: linux-x64 (Alpine Linux - 3.17.4)
[2023-08-29T09:19:14Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT Cypress Version: 10.10.0
```

and 2 seconds later on a different build agent:

```
[2023-08-29T09:21:07Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT The Cypress App could not be downloaded.
[2023-08-29T09:21:07Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT
[2023-08-29T09:21:07Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT Does your workplace require a proxy to be used to access the Internet? If so, you must configure the HTTP_PROXY environment variable before downloading Cypress. Read more: https://on.cypress.io/proxy-configuration
[2023-08-29T09:21:07Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT
[2023-08-29T09:21:07Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT Otherwise, please check network connectivity and try again:
[2023-08-29T09:21:07Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT
[2023-08-29T09:21:07Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT ----------
[2023-08-29T09:21:07Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT
[2023-08-29T09:21:07Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT URL: https://download.cypress.io/desktop/10.10.0?platform=linux&arch=x64
[2023-08-29T09:21:07Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT Error: Failed downloading the Cypress binary.
[2023-08-29T09:21:07Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT Response code: 502
[2023-08-29T09:21:07Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT Response message: Bad Gateway
[2023-08-29T09:21:07Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT
[2023-08-29T09:21:07Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT ----------
[2023-08-29T09:21:07Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT
[2023-08-29T09:21:07Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT Platform: linux-x64 (Alpine Linux - 3.17.4)
[2023-08-29T09:21:07Z] ➤ YN0000: │ cypress@npm:10.10.0 STDOUT Cypress Version: 10.10.0
```

**Comment by MikeMcC399** [comment]
@bradleyayers 

https://www.cloudflarestatus.com/#past-incidents is showing some related issues. 

> Investigating - Cloudflare is investigating issues with elevated  5xx /499s errors for some customers.

Is it still happening for you?

**Comment by bradleyayers** [comment]
Nope it was very transient blip, there were a bunch of agents running that _did_ succeed, it was just couple that failed. It's working fine now. We experience the flakiness often enough for it to be painful (personally I would see it every week), so I'm trying to work out whether we should cache it on our side somewhere.

**Comment by MikeMcC399** [comment]
@bradleyayers 

It may depend also on your location due to the distributed nature of the Cloudflare infrastructure. FWIW https://www.cypressstatus.com/ shows no outage.

Good to hear that it is working fine now. That would tie in with the Cloudflare notice which says that the incident has been resolved.

**Comment by johndevedu** [comment]
Our infra is hitting this issue several times last couple of days...

```
[2023-09-22T04:00:49.850Z] #18 304.6 npm ERR! URL: https://download.cypress.io/desktop/4.12.1?platform=linux&arch=x64
[2023-09-22T04:00:49.850Z] #18 304.6 npm ERR! Error: Failed downloading the Cypress binary.
[2023-09-22T04:00:49.850Z] #18 304.6 npm ERR! Response code: 502
[2023-09-22T04:00:49.850Z] #18 304.6 npm ERR! Response message: Bad Gateway
```
Is there a way to configure the postinstall to look to a different url? i.e. internal mirror repo.

**Comment by MikeMcC399** [comment]
@johndevedu 

> Is there a way to configure the postinstall to look to a different url? i.e. internal mirror repo.

Yes, there is a whole section on this in [Advanced Installation](https://docs.cypress.io/guides/references/advanced-installation).

(Cypress `4.12.1` is more than 3 years old, so perhaps you should be considering updating sometime?)

**Comment by MatthewMaclean** [comment]
We see this periodically in our CICD as well:
```
Fri, 10 Nov 2023 14:46:15 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDOUT Installing Cypress (version: 12.5.1)
Fri, 10 Nov 2023 14:46:15 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDOUT 
Fri, 10 Nov 2023 14:46:15 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDOUT [STARTED] Task without title.
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDERR [FAILED] The Cypress App could not be downloaded.
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDERR [FAILED] 
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDERR [FAILED] Does your workplace require a proxy to be used to access the Internet? If so, you must configure the HTTP_PROXY environment variable before downloading Cypress. Read more: https://on.cypress.io/proxy-configuration
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDERR [FAILED] 
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDERR [FAILED] Otherwise, please check network connectivity and try again:
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDERR [FAILED] 
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDERR [FAILED] ----------
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDERR [FAILED] 
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDERR [FAILED] URL: https://download.cypress.io/desktop/12.5.1?platform=linux&arch=x64
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDERR [FAILED] Error: Failed downloading the Cypress binary.
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDERR [FAILED] Response code: 502
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDERR [FAILED] Response message: Bad Gateway
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDERR [FAILED] 
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDERR [FAILED] ----------
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDERR [FAILED] 
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDERR [FAILED] Platform: linux-x64 (Ubuntu - 22.04)
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDERR [FAILED] Cypress Version: 12.5.1
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDOUT The Cypress App could not be downloaded.
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDOUT 
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDOUT Does your workplace require a proxy to be used to access the Internet? If so, you must configure the HTTP_PROXY environment variable before downloading Cypress. Read more: https://on.cypress.io/proxy-configuration
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDOUT 
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDOUT Otherwise, please check network connectivity and try again:
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDOUT 
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDOUT ----------
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDOUT 
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDOUT URL: https://download.cypress.io/desktop/12.5.1?platform=linux&arch=x64
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDOUT Error: Failed downloading the Cypress binary.
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDOUT Response code: 502
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDOUT Response message: Bad Gateway
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDOUT 
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDOUT ----------
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDOUT 
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDOUT Platform: linux-x64 (Ubuntu - 22.04)
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0000: │ cypress@npm:12.5.1 STDOUT Cypress Version: 12.5.1
Fri, 10 Nov 2023 14:46:18 GMT   ➤ YN0009: │ cypress@npm:12.5.1 couldn't be built successfully (exit code 1, logs can be found here: /tmp/xfs-e5f33eb8/build.log)
```

I've disabled Cypress install in GitHub CICD jobs that don't require the binary to be installed, but it's still annoying to have CICD failures due to the post-install step failing. Can Cypress add some post-install retries for this style of error? Yarn does it for installs, but they don't cover this type of failure. In our case, we have 5 different jobs spinning up in parallel and running Cypress tests. Occasionally 1 of the 5 will fail, which indicates it's flaky and short-lived.

**Comment by jasonplumb** [comment]
Facing the same issue also - do we need to use a different version?

**Comment by MikeMcC399** [comment]
@Jay-Plumb 

> Facing the same issue also - do we need to use a different version?

This is an infrastructure issue, which is independent of the Cypress version. It will usually recover on its own if you wait. Which Cypress version are you using?

**Comment by jasonplumb** [comment]
@MikeMcC399 I was on 12.6.0 - i've noticed the problem several times over the last two weeks. Thanks for letting me know it's an infrastructure issue

**Comment by martin-nef** [comment]
Still regularly getting this, on version 13.8.1

**Comment by MikeMcC399** [comment]
@martin-nef 

> Still regularly getting this, on version 13.8.1

It would probably be worth opening a new issue if this is still occurring. You should give some details about where the 502 Bad Gateway error is occurring:

- CI Provider? If yes, which one e.g. GitHub Actions
- If CI, is the system runner provided by the CI provider or is it locally hosted? e.g. GitHub-hosted runner / self-hosted runner
- If locally hosted or similar, where is the system located? (See https://www.cloudflarestatus.com/ for list of relevant geographies and localities).
